crates/typst/src/layout/grid/layout.rs

```diff
@@ -358,20 +590,60 @@ impl CellGrid {
                         &fill.resolve(engine, x, y)?,
                         align.resolve(engine, x, y)?,
                         inset,
+                        stroke.resolve(engine, styles, x, y)?,
                         styles,
                     );
                     Ok(Entry::Cell(new_cell))
                 }
             })
             .collect::<SourceResult<Vec<Entry>>>()?;
 
-        Ok(Self::new_internal(tracks, gutter, resolved_cells))
+        let row_amount = resolved_cells.len().div_ceil(c);
+        for (span, line) in pending_hlines {
+            let line = {
+                let Line { index: y, start, end, stroke, position } = line;
+                if position == LinePosition::After && (!has_gutter || y + 1 == row_amount)
+                {
+                    // Just place the line on top of the next row if
+                    // there's no gutter and the line should be placed
+                    // after the one with given index.
+                    // Note that placing after the last row is also the same as
+                    // just placing on the grid's bottom border, even with
+                    // gutter.
+                    Line {
+                        index: y + 1,
+                        start,
+                        end,
+                        stroke,
+                        position: LinePosition::Before,
+                    }
+                } else {
+                    Line { index: y, start, end, stroke, position }
+                }
+            };
+            let y = line.index;
+            // There must exist at least one more row for this line to be
+            // actually drawn if the "After" position was chosen.
+            let necessary_rows =
+                if line.position == LinePosition::After { y + 1 } else { y };
+            if row_amount < necessary_rows {
+                bail!(span, "cannot place horizontal line at invalid row");
+            }
+            if hlines.len() <= y {
+                hlines.resize_with(y + 1, Vec::new);
+            }
+            hlines[y].push(line);
```

Comment:
One possible optimization here is to keep track of the maximum `y` given for hlines during the grid item loop. Then, we would only have to check this kind of violation once, and also only have to resize once. However, it would mean that we would error at the last "bad" hline, not at the first. What would be preferable here?

Reply:
Ok. I think it should be fine for now as well; we can always change our mind later on this.